src/IO/WriteBufferFromS3.cpp

```diff
@@ -94,41 +105,56 @@ WriteBufferFromS3::WriteBufferFromS3(
 
 void WriteBufferFromS3::nextImpl()
 {
+    LOG_TRACE(log, "Incoming data size {}, memory buffer size {}", offset(), memory.size());
+
     if (!offset())
         return;
 
-    /// Buffer in a bad state after exception
-    if (temporary_buffer->tellp() == -1)
-        allocateBuffer();
+    chassert(memory.size() >= offset());
 
-    size_t size = offset();
-    temporary_buffer->write(working_buffer.begin(), size);
-
-    ProfileEvents::increment(ProfileEvents::WriteBufferFromS3Bytes, offset());
-    last_part_size += offset();
-    if (write_settings.remote_throttler)
-        write_settings.remote_throttler->add(offset());
-
-    /// Data size exceeds singlepart upload threshold, need to use multipart upload.
-    if (multipart_upload_id.empty() && last_part_size > settings.max_single_part_upload_size)
+    if (multipart_upload_id.empty())
+    {
         createMultipartUpload();
+    }
 
-    chassert(upload_part_size > 0);
-    if (!multipart_upload_id.empty() && last_part_size > upload_part_size)
+    if (!multipart_upload_id.empty())
     {
-        writePart();
-
-        allocateBuffer();
+        if (offset() == memory.size())
+            writePart();
```

Comment:
It looks like we expect here that `nextImpl` is called only when the buffer is full. I suspect that it is not always true (this is not an invariant that we check/support for buffers).

Reply:
It just has to be true for such implementation. If it is not than there is no way to implement one-copy buffer. As long as we need to control parts sizes we have to either do double-copy or demand that nextImpl is call when buffer is full (last part is ok to be lesser).

Reply:
But you are right in that thing: there is happens call next which is supposed to flush the buffer. And WriteBufferFromS3 has to flush it due to realisation WriteBuffer::next. WriteBuffer::next just rewind the buffer what ever what happens in successor. And this is happening indeed, I see "DB::Exception: Failed to complete multipart upload. No parts have uploaded." in CI.
I'm thinking what to do with that case.

Reply:
So, I find the way to address it.
I don't have much choice when next is called. I prefer to flush the buffer and make a new part with current partially filled buffer. That could lead to creating parts with sizes smaller than 5M.  If it happens in production enviroment S3 returns an error. I suppose that we do not do such thing. I added assert statement in order to check it with debug build. Any way, all builds would have an LOG_WARNING message for such case.